metagraph/src/annotation/row_diff_builder.cpp

```diff
@@ -378,34 +397,35 @@ void build_pred_succ(const std::string &graph_fname,
         std::vector<bool> pred_boundary_buf;
 
         for (uint64_t i = start; i < std::min(start + BS, graph.num_nodes() + 1); ++i) {
-            BOSS::edge_index boss_idx = graph.kmer_to_boss_index(i);
-            if (!dummy[boss_idx]) {
-                BOSS::edge_index next = boss.fwd(boss_idx);
-                assert(next);
-                if (!dummy[next]) {
-                    while (rd_succ.size() && !rd_succ[next]) {
-                        next--;
-                        assert(!boss.get_last(next));
-                    }
-                    succ_buf.push_back(to_row(graph.boss_to_kmer_index(next)));
+            bool find_succ = true;
+            if (succinct) { // Legacy code for DBGSuccinct
+                const BOSS &boss = succinct->get_boss();
+                BOSS::edge_index boss_idx = succinct->kmer_to_boss_index(i);
+                    BOSS::edge_index next = boss.fwd(boss_idx);
+                    assert(next);
+                if ((*dummy)[next]) {
+                    find_succ = false;
+                        }
+            }
+            auto with_rd_succ = [&](bit_vector const& rd_succ) {
```

Comment:
This seems to fix `RowDiff` test group. But generally I'm not sure if it's really OK for `route_at_forks` to just return an empty bit vector... Should I change it to return optional to indicate that it sometimes fails?

Reply:
Or maybe make `route_at_forks` return `shared_ptr` too, which will be `rd_succ` when possible and `get_last()` otherwise. Should simplify some branching in the code.